Implement tooling support to connect using custom IPC channels.

lateralusX opened the pull request:
dsrouter presents same default IPC channel to tools as a regular coreclr process so there is a risk of collision with its own coreclr runtime default IPC channel (if used).

dsrouter supports custom named IPC channels using its -ipcs argument, but current tooling can't connect to custom named IPC channels, only supporting the default IPC channel through use of --process-id or --name arguments.

This PR extend tooling support in `dotnet-counters` and `dotnet-trace` extending the `--diagnostic-port` command accepting custom IPC channels to connect to, instead of default channels based on process id.

lateralusX commented:
//CC @josalem

jander-msft commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -201,15 +201,23 @@ private static bool TryGetDefaultAddress(int pid, out string defaultAddress)
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
             {
-                defaultAddress = $"dotnet-diagnostic-{pid}";
+                if (File.Exists($"{IpcRootPath}dotnet-dsrouter-diagnostic-{pid}"))
```
These feel like an over-optimization for dsrouter scenarios that impacts all diagnostics scenarios. In my opinion, all diagnostic tools should not have to scan two addresses spaces when discovering processes to diagnose.

This also has the issue of if dsrouter leaves behind its custom diagnostic port Unix domain socket file, then that process ID is forever inaccessible until the file is cleared out. For example, dsrouter runs as process ID 12345, then is killed (crashes, etc) leaving behind the file, then another non-dsrouter process starts as process ID 12345; this process cannot be discovered due to the proposed algorithm.

It might be better to leave this alone and then add an ability to the diagnostic tools to reach out a specific address for connection. For example, the tools can be provided the port name and the mode in which they connect e.g. `dotnet-counters --dignostic-port dotnet-dsrouter-diagnostic-{pid} --diagnostic-port-mode connect`. I know that this doesn't work well for Linux given the port has an "arbitrary" disambiguator, but I'm proposing a general concept rather than a specific implementation.

lateralusX replied:
yes, part of this PR is to have the discussion, @josalem and I discussed the ability to open up to explicit use named ipc ports in all tooling, dsrouter can already do that using -ipcs argument, so if we are ok adding the argument I'm perfectly happy to do that way instead. Just one side effect, then dsrouter will have issues using the --process-id/--name argument since it won't bind the default port, but we could always use the option to disable diagnostics in dsrouter process and detect that case and warn as a fallback.

lateralusX replied:
If we are OK doing that, I can adjust this PR tomorrow.

noahfalk replied:
My 2 cents ...
I think we are better off making the dsrouter proxy connection something that a tool user configures explicitly rather than trying to trying to broaden the meaning of the --process-id/--name arguments. For example the workflow might look like this:
```
# Run the proxy
> dsrouter some_args
Proxy is listening for connection at address: /tmp/android_proxy.sock
To connect local dotnet tools use argument: --diagnostic-port /tmp/android_proxy.sock,connect

# Run a tool
> dotnet-trace --diagnostic-port /tmp/android_proxy.sock,connect ...
```

lateralusX replied:
@noahfalk Thanks for feedback, please check my long comment below, but I'm landing on something similar, either having a explicit --connect option that can go with --diagnostic-port, making it explicit to connect to a custom named IPC channel. Not sure if we preferer the runtime way backing it all into the --diagnostic-port argument or having it as an explicit option. So either we do it like you propose, `--diagnostic-port /tmp/android_proxy.sock,connect` or we do it, `--diagnostic-port /tmp/android_proxy.sock --connect`.

noahfalk replied:
Another variation to consider `--connect /tmp/android_proxy.sock`. Personally I like it because it is shorter but I know these can be very subjective so I'd love to get some other opinions from you guys.

lateralusX replied:
So then we reserve the --diagnostic-port for reverse connect, and the new `--connect` to connect against a custom named IPC channel, I'm fine with that.

jander-msft commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouterFactory.cs:
```diff
@@ -478,7 +487,17 @@ static string GetDefaultIpcServerPath()
+                {
+                    logger?.LogWarning($"Default IPC server path, {Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}-*-socket")}, already in use. To disable default diagnostics for dotnet-dsrouter, set COMPlus_EnableDiagnostics=0 and re-run.");
+
+                    path = Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-dsrouter-diagnostic-{processId}-{(long)diff.TotalSeconds}-socket");
```
nit: Given that this is not a diagnostic port to diagnose dsrouter, could you change the name so that it is confused with being the diagnostic port for dsrouter e.g. `dotnet-dsrouter-{processId}`?

lateralusX replied:
jupp

lateralusX replied:
but if we implement connecting to named ports we don't need to fallback but can fail with error informing user to either turn of default IPC diagnostic channel or use -ipcs for an explicit named IPC channel and use --diagnostic-port and diagnostic-port-mode connect, if we decide to move down that road.

jander-msft replied:
The only way to turn off the diagnostic port server in the runtime is to set ComPlus_EnableDiagnostics=0. This is problematic for two reasons:

1. It's difficult (error prone?) to tell all users of the tool to set this environment variable for only this executable every single time you want to launch it.
1. It disables the ability to debug the process. So while you may be able to run it, you won't be able to debug it.

Ideally, there's a separate environment variable for turning off just the diagnostic port services or, even better, a runtime config switch that can be specified in the .runtimeconfig.json file that is deployed with the app.

lateralusX replied:
There is already other scenarios where you will need to run with env variables (like reverse connect), but I don't like it either since it complicates things, but if its the only option we have to disable it, that is what we have and it works on older .net versions as well. I would also love to put this in `.runtimeconfig.json`, but none of the diagnostics env's are currently included, don't know if env's in that file are set to late (needs to be set by the host or very early when launching runtime), if that's the case, then `.runtimeconfig.json` can't hold these settings.

Having a good way to turn of the default IPC listener in CoreCLR/MonoVM would be a good thing, but if we do that, we will get a dependecy on a specific .net runtime version as well when running dsrouter (but not other diagnostic tools) and that is sub optimal as well.

So sounds like we are out of good options on this, and the whole motivation to replace the default IPC channel when running dsrouter in server-server or server-client mode (client-server is unaffected by this), was to make it transparent for end users to connect and trace remote targets handled by dsrouter.

noahfalk replied:
> the whole motivation to replace the default IPC channel when running dsrouter in server-server or server-client mode (client-server is unaffected by this), was to make it transparent for end users to connect and trace remote targets handled by dsrouter.

IMO the transparency winds up being confusing because it makes it ambiguous. Is the tool diagnosing dsrouter itself (it is a .NET Core app after all) or is the tool being redirected through the proxy and diagnosing a remote process? It feels akin to using file system commands on symbolic links. For someone like myself that doesn't do it frequently I'd have to go find some docs to determine which commands indirect through the link and which ones don't.

Assuming that the underlying goal for transparency was ease-of-use I think we have alternative UX options that will remain easy without attempting to be transparent, such as [using --diagnostic-port](https://github.com/dotnet/diagnostics/pull/2374#discussion_r654223937) or perhaps some other explicit command-line option.

lateralusX commented:
Thanks @jander-msft for all feedback so far!

Feels like I need to rebase thinking a little. Motivation of "replacing" the default IPC channel in dsrouter was to make it transparent to users running diagnostic tooling with --process-id or --name, and tooling would take care of getting all traffic over to mobile device (or some other container or remote running app).

dsrouter can run using a custom IPC channel using its -ipcs argument, but currently diagnostic tooling like dotnet-trace can't connect to anything expect the default IPC channel.

We could always rely on -ipcs and add option in diagnostic tooling to connect towards that channel, but that "complicates" the default scenario for dsrouter since it adds additional config both when launching dsrouter as well as running diagnostic tooling.

Since it looks like we can't do any decent heuristics in diagnostic tooling to connect to a different named IPC channel if it exists (due to potential error cases), maybe we should add a simple way to inform diagnostic tooling that user intends to connect to dsrouter's default IPC channel  instead of runtimes default IPC channel in targeted process. If we make that argument "light" it won't be to much for the user to type, possible drawback/error case, user misses to include argument and start to trace dsrouter instead of remote target and won't see any difference, so that not optimal either.

So that brings us back to a situation where the tooling needs to know that it's going to connect to dsrouter and pick the right default IPC channel to do the job, and doing that automatically would have simplified things and reduced errors.

So based on all above, this is where I'm currently at.

Open up so diagnostic tooling can connect "custom" named IPC channels, not just the default. We already have --diagnostic-port for the reverse connect scenario, so we will need something additional telling tooling to connect using the --diagnostic-port and not setup a reverse diagnostic server in the tool, to keep arguments short, maybe we should just have a --connect option to connect instead of listen. This should be straight forward and a generic extension to all tooling.

Teach tooling about dsrouter and that the default IPC channel is named differently, since dsrouter is an important component in the diagnostics toolset, I believe it is OK to include some specific knowledge about it into the tools. Maybe a more "safe" approach could include the following heuristics in TryGetDefaultAddress:

* Check name of process, if its dotnet-dsrouter* check for dsrouter IPC channel.
* If both above checks fail, fallback to default IPC channel.

This will only get us into looking for a different named IPC channel if the process is named dotnet-dsrouter*. This won't work if dsrouter was executed using dotnet run, but that is not how the tooling gets installed by end users and running tooling using dotnet run already breaks other process discovery scenarios, like using --name, so think it could be a decent approach.

If heuristics fails there is always the option to start dsrouter using -ipcs and then pass --diagnostic-port and --connect arguments to tooling, but that should only be needed in cases where dsrouter is not run using dotnet-dsrouter command.

I draft that up in this PR and we can take discussion from there.

noahfalk commented:
> Teach tooling about dsrouter and that the default IPC channel is named differently

Is the goal behind this to avoid users unintentionally tracing the proxy app when they meant to trace the app being proxied? What if we left the default IPC channel alone and made the tools give a warning like this when they recognized the process name:
```
> dotnet-trace --processId 1234
ProcessId 1234 is a known proxy 'dotnet-dsrouter'.
If you intended to connect to the app being proxied use the --connect argument instead with the
proxy's advertised address.
If you intended to diagnose the proxy itself, re-run this command with --force to ignore this warning.
```

> If you user want to connect towards a custom IPC channel, then use --diagnostic-port mycustomchannel --connect.

What if we defined that `--connect mycustomchannel` is the more compact way of expressing the same thing? Do folks think users would find this more confusing?

> possible drawback/error case, user misses to include argument and start to trace dsrouter instead of remote target and won't see any difference, so that not optimal either

Hopefully we can avoid having one command be a subset of the other. For example no forgotten argument or typo would convert `dotnet-trace --connect mycustomchannel` into `dotnet-trace --processId 1234`. Certainly confusing the two scenarios remains a possibility but I think we've got options to help people avoid that mistake if it proved to be a problem.


> Open up so diagnostic tooling can connect "custom" named IPC channels

This sounds good to me. Even without dsrouter I expect it was only a matter of time. For example it would be useful to tunnel across container boundaries.

lateralusX commented:
> Is the goal behind this to avoid users unintentionally tracing the proxy app when they meant to trace the app being proxied? What if we left the default IPC channel alone and made the tools give a warning like this when they recognized the process name:
> 
> ```
> > dotnet-trace --processId 1234
> ProcessId 1234 is a known proxy 'dotnet-dsrouter'.
> If you intended to connect to the app being proxied use the --connect argument instead with the
> proxy's advertised address.
> If you intended to diagnose the proxy itself, re-run this command with --force to ignore this warning.
> ```

Yeah, sounds reasonable, was hoping we could have this scenario a little more transparent to the end users, but that in turn opens up a range of other problems, so a warning above would at least make sure users are doing the right thing and since dsrouter already have a default or custom IPC channel printed on startup, that could be used with --connect argument.

josalem commented:
I think Noah's points are all valid. I also think having too many ways of specifying a connection point might end up getting confusing, though. I would hope we can keep the tools from having 1:1 connection specifying flags to runtime connection options. Perhaps we should look at unifying things under a specific flag with metadata on it?

I'm imagining something like:
```sh
# default patterns
dotnet trace collect -p <pid>
dotnet trace collect -n <app name>

# more complex patterns
dotnet trace collect --address tcp://localhost:1234,connect
dotnet trace collect --address tcp://localhost:1234,listen
dotnet trace collect --address uds:///Users/joshmo/myunixdomain.sock,connect
dotnet trace collect --address uds:///Users/joshmo/myunixdomain.sock,listen
dotnet trace collect --address namedpipe://./pipes/mynamedpipe,connect
dotnet trace collect --address namedpipe://./pipes/mynamedpipe,listen
```

This gives us a framework for adding more complex connection scenarios without necessarily adding new flags or more complex user invocations at the cost of a slightly more complex config parsing phase on our part. We could add as many tags to the end of the address as we want without necessarily increasing the cognitive load of specifying a simple scenario as long as our defaults are reasonable. We get the added benefit of the address syntax matching the syntax for the `DOTNET_DiagnosticPorts` as well. We could also reuse the `--diagnostic-port` flag and just update the way it is used. I understand the case for not making a breaking change to usage patterns, but we could accept both forms for a version or two and print a warning about the behavior changing.

CC @shirhatti in case you have thoughts from a user E2E perspective.

lateralusX commented:
I had similar thoughts combining this into the same argument, like how our DOTNET_DiagnosticPorts work, but was a little afraid that the user experience would suffer making it to complex, if we are going to combine it maybe we should just use the one we already have diagnostic-port and just add connect argument to it, listen will be default to keep current default behaviour.

noahfalk commented:
> I'm imagining something like:

Were you intentionally changing --diagnostic-port to --address? It felt like we could do the same thing using the argument we already have and there would be no need for anyone to learn a new one?

```
# default patterns
dotnet trace collect -p <pid>
dotnet trace collect -n <app name>

# more complex patterns
dotnet trace collect --diagnostic-port tcp://localhost:1234,connect
dotnet trace collect --diagnostic-port tcp://localhost:1234,listen
dotnet trace collect --diagnostic-port uds:///Users/joshmo/myunixdomain.sock,connect
dotnet trace collect --diagnostic-port uds:///Users/joshmo/myunixdomain.sock,listen
dotnet trace collect --diagnostic-port namedpipe://./pipes/mynamedpipe,connect
dotnet trace collect --diagnostic-port namedpipe://./pipes/mynamedpipe,listen

# I assume if we used --diagnostic-port we would still support the shortened address forms accepted elsewhere?
dotnet trace collect --diagnostic-port localhost:1234
dotnet trace collect --diagnostic-port /Users/joshmo/myunixdomain.sock
dotnet trace collect --diagnostic-port //./pipes/mynamedpipe
```

lateralusX commented:
I also believe it needs to have good defaults, so users don't need complex address formats.

josalem commented:
> Were you intentionally changing --diagnostic-port to --address?

When I wrote that I was trying to avoid a breaking change to how `--diagnostic-port` is used, but I'm also okay reusing that flag as well. It seems perfectly reasonable to me.

>  I assume if we used --diagnostic-port we would still support the shortened address forms accepted elsewhere?

That's how I imagine it as well.

We would make assumptions based on platform and have reasonable default values. e.g., on Windows we would assume a schema of `namedpipe://` and on non-Windows we would assume a schema of `uds://`, etc. Same goes for tags, e.g., we would assume `connect` if unspecified to preserve existing behavior of the tools.

jander-msft commented:
> Same goes for tags, e.g., we would assume `connect` if unspecified to preserve existing behavior of the tools.

You mean listen (establishes its own server to listen for runtimes to connect to it) is the default, correct? I know that dotnet-monitor implicitly uses listen when using --diagnostic-port, and I believe dotnet-counters/trace do the same.

josalem commented:
ah, yes, you're right. The default for the `--diagnostic-port` flag would be the `listen` tag since using that flag would make the tool start a Diagnostics IPC server.

lateralusX commented:
Sound like we are inline with changes needed. I will adjust this PR accordingly.

lateralusX commented:
First set of changes done, I adjusted the tools that already supported `--diagnostic-port` argument, `dotnet-trace` and `dotnet-counters` to handle extended format, I also added `dotnet-stack` into the mix, other tools (like dotnet-gcdump, dotnet-dump) are still not touched since they didn't support `--diagnostic-port` in first place (and we don't currently support them on MonoVM).

I didn't add any protocol parsing logic to reduce the initial changes, so right now no protocol prefix is handled and you only use the default protocol for the platform (namedpipe or uds).

The default for `--diagnostic-port`, if nothing else is specified is `listen` mode, to be backward compatible, and I added support for `connect` mode. Parsing logic is inline with parsing of `DOTNET_DiagnosticPorts` in runtime, so:

`--diagnostic-port mycustomchannel,connect`

will tell the tool to connect to a IPC channel called mycustomchannel.

and if you just run like before:

`--diagnostic-port mycustomchannel`

means tool will setup a reverse IPC server and listen for incoming connections, indentical to how `--diagnostic-port` works today and is equivalent to:

`--diagnostic-port mycustomchannel,listen`

jander-msft commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
+            {
+                var parts = config.Split(',');
+
+                Address = parts[0];
```
Should validate that the address is not empty.

jander-msft commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
         public abstract Task WaitForConnectionAsync(CancellationToken token);
     }
 
+    internal class IpcEndpointConfig
```
This should be moved to its own file. It has no relationship to IpcEndpoint and its derived classes.

lateralusX replied:
Didn't think it was that off since its the configuration used together with IpcEndpoint's, but could of course be move to a different file.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
+        public string Address { get; } = "";
+        public PortType Type { get; } = PortType.None;
+
+        public IpcEndpointConfig(string config)
```
The None value was added to make it possible to test that you don't have a valid address, but in theory you can do the same checking for null or empty address property but felt less obvious, maybe it would be better to have properties to check what type it has instead of accessing the underlying enum, like isListen and isConnect and in case when we have address null or empty string both will return false. I write something up.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
 
+    internal class IpcEndpointConfig
+    {
+        internal enum PortType
```
Wasn't sure we wanted to include this in any public API at this point since it's mainly for our internal tools, maybe we should start with internal and then if there is a need, switch to public?

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
+                Address = parts[0];
+                Type = PortType.Listen;
+
+                for (int i = 1; i < parts.Length; i++)
```
Initially I added the same keywords that we have in runtime (also parse them using a loop), so it was more than 1, but we could of course just get rid of the loop as long as we only have one keyword and reintroduce the loop if we need more in the future.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
+
+                for (int i = 1; i < parts.Length; i++)
+                {
+                    if (string.Compare(parts[i], "connect", StringComparison.OrdinalIgnoreCase) == 0)
```
What's the benefit, except comparing against bool instead of == 0?

jander-msft replied:
Depends on the compare implementation, but compare doesn't have to just stop at the first differing character; it could compute some value for both strings and then compare them. Whereas, equality comparing will check character-by-character and immediately stop after the first difference.

lateralusX replied:
When you do case sensitive compare (StringComparison.Ordinal),  equals can do optimized check, BCL implements it as:

```
SpanHelpers.SequenceEqual(
    ref Unsafe.As<char, byte>(ref strA.GetRawStringData()),
    ref Unsafe.As<char, byte>(ref strB.GetRawStringData()),
    ((uint)strA.Length) * sizeof(char));
```

but when you do a case insensitive compare the implementation depends on string content and in some scenarios both implementations get very close or pretty much identical, but the fact that the Equals is not required to report different results (just bool) while Compare needs to support sorting order puts Equals in position to accept more implementation optimizations.

Looking in codebases like mono/mono and .net framework reference source those tend to favour string.Compare for case insensitive compare, while dotnet/runtime seems to favour string.Equals. So seems to shifted in favour of string.Equals over time, interesting!

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,88 @@ internal abstract class IpcEndpoint
+            }
+        }
+
+        public static Stream Connect(string address, TimeSpan timeout)
```
Yes couldn't find a good spot for them in existing classes, but we can of course add a new class for just these methods, like the one proposed.

lateralusX commented:
Added schema parsing rules to `--diagnostic-port`:

Default will be platform specific namedpipe or uds, so not needed to add schema prefix to path, but if done, it needs to follow regular URI syntax as defined by https://datatracker.ietf.org/doc/html/rfc3986 and implemented by System.Uri.

The following is accepted:

**UDS**
myport => myport
uds:myport => myport
/User/mrx/myport.sock => /User/mrx/myport.sock
uds:/User/mrx/myport.sock => /User/mrx/myport.sock
uds://authority/User/mrx/myport.sock => /User/mrx/myport.sock

**NamedPipe**
myport => myport
namedpipe:myport => myport
\\.\pipe\myport => myport (dropping \\.\pipe\ is inline with implemented namedpipe client/server)
namedpipe://./pipe/myport => myport (dropping authority and /pipe/ is inline with implemented namedpipe client/server)
namedpipe:/pipe/myport  => myport (dropping /pipe/ is inline with implemented namedpipe client/server)
namedpipe://autohrity/myport => myport


NOTE, using schema:// is only used when having an authority (machine:port) as part of the URI, for namedpipe . is considered localhost, but nothing similar for uds. Since the authority will be dropped for namedpipe and uds, they could be anything

noahfalk commented on src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs:
```diff
@@ -149,47 +151,57 @@ private static async Task<int> Collect(CancellationToken ct, IConsole console, i
+                Process process = null;
                 DiagnosticsClientBuilder builder = new DiagnosticsClientBuilder("dotnet-trace", 10);
-                bool shouldResumeRuntime = ProcessLauncher.Launcher.HasChildProc || !string.IsNullOrEmpty(diagnosticPort) || resumeRuntime;
+                bool shouldResumeRuntime = ProcessLauncher.Launcher.HasChildProc || portConfig.IsListenConfig || resumeRuntime;
```
If someone explicitly writes --resume-runtime: false on the command-line I think they would be very confused if the tool still sends the resume command. I think we should either honor the false setting or give an error saying that resumeRuntime:false is not allowed in a given situation.

lateralusX replied:
Jupp though about that as well, we now potentially have some arguments that would even out each other. I think we should honor --resume-runtime: false in cases where we run child process or is setup in reverse connect scenario, since that is what the user explicitly asked.

lateralusX replied:
I implemented that approach in this PR:
shouldResumeRuntime = (ProcessLauncher.Launcher.HasChildProc || portConfig.IsListenConfig) && resumeRuntime;
OK?

noahfalk replied:
From our previous discussion didn't we want to send the resume command when connecting? I thought we wanted support scenarios like this:
```
# launch app
export DOTNET_DiagnosticPorts=foo.sock,listen,suspend
myapp.exe

# launch tool and resume
dotnet-trace --diagnostic-port foo.sock,connect ...
```

I was thinking the policy could be really simple:
```
shouldResumeRuntime = resumeRuntime;
```

lateralusX replied:
Agree, lets get rid of the complete expression, intent should be to always resume runtime, regardless if we listen,connect and user can override by passing --resume-runtime: false, regardless how you are running.

noahfalk commented on src/Tools/dotnet-stack/ReportCommand.cs:
```diff
@@ -33,18 +33,37 @@ internal static class ReportCommandHandler
+
+                    if (!portConfig.IsConnectConfig)
+                    {
+                        Console.WriteLine("--diagnostic-port only supports connect mode.");
```
Is this limitation to reduce the immediate implementation cost or for something else? I worry it will confuse users if our tools aren't handling this diagnostic-port argument uniformly. If we wanted to save implementation cost I'd propose not supporting it at all for now.

lateralusX replied:
Yes, dotnet-stack didn't support --diagnostic-port and I tried to reduce all changes needed and would still be nice to get dotnet-stack working with dsrouter in server-server, server-client mode. If we don't support it we won't be able to use dotnet-stack with dsrouter, so I believed it was better to at least add something than nothing.

noahfalk replied:
My personal preference would be to support it fully if it is important or leave it unsupported for now if not. However if you think it is important to get the partial solution quickly then I'd suggest we should at least change the error message to make it clear that it is dotnet-stack specifically that doesn't support this, not a general limitation in --diagnostic-port across our tools. Perhaps:
```
Console.WriteLine("dotnet-stack support for a listening mode diagnostic-port is not yet available");
```

Aside: If adding listen support to dotnet-stack is hard I think it means we need to improve how we've factored code for the tools. Ideally the code that supports all these different connection options should be somewhere that we share it easily.

lateralusX replied:
The logic is a little spread out, and each piece I touch tends to call for more refactoring, so a little afraid that this PR will just continue to bloat :), but I can of course take a look and see (don't think its that much), but will be some duplication coming in from dotnet-trace or dotnet-counters. The other tools have the whole process execution feature connected to the --diagnostic port as well, but maybe we could at least leave that outside for now and focus only on --diagnostic-port argument?

noahfalk replied:
> so a little afraid that this PR will just continue to bloat :)

Two options if you want to press the merge button sooner is to leave the functionality out for a later PR or only change the error message. While I error message only isn't my preference, I don't begrudge you if that is what you decide is the best course : )

lateralusX replied:
After discussing and thinking about it a little more, maybe we should just leave it out in this PR and then roll a new PR implementing support for --diagnostic-port in dotnet-stack.

lateralusX replied:
Removed it from this PR.

noahfalk commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs:
```diff
@@ -26,6 +26,11 @@ public sealed class DiagnosticsClient
         {
         }
 
+        public DiagnosticsClient(string address) :
```
internal for now? I've been assuming that we'll want to look over a bunch of the reverse server and endpoint work, then make an explicit decision about we feel comfortable marking public.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,126 @@
+
+        public string Address { get; }
+
+        public bool IsConnectConfig {
```
Not really, but since we handle diagnosticPort as an empty string/null, being able to parse that and continue getting the address in places where we use it reduce the changes needed in the tooling code, so being amble to create an instance even when diagnosticPort was empty/null simplified that. I guess we could do more major changes to the whole process-id, name, diagnosticPort handling in the tooling, but maybe we should do that separately.

noahfalk replied:
> so being able to create an instance even when diagnosticPort was empty/null simplified that

OK, as long as it is useful I'm happy : ) Should the time come that we think this deserves to be part of the public API I'd step my scrutiny a good bit.

lateralusX replied:
Do you think it would be better to let the caller handle case of failure and then passing around a null ref through its methods to handle case where diagnostic port is empty/null? That change is simple to be done, making this class cleaner and we put the burden on the caller to better handle it? Would that pass your bar? :)

noahfalk replied:
> Do you think it would be better to let the caller handle case of failure and then passing around a null ref through its methods to handle case where diagnostic port is empty/null, like in DiagnosticsClientBuilder.Build?

I was speculating that it would be better but I hadn't gone to look at every usage and contemplate what the alternative code was going to look like. My assumption is that the callers already need to handle other exceptions that might be thrown from Parse() so presumably a FormatException saying address is empty has the same control flow needs as an exception saying the scheme is bad.

lateralusX replied:
Dropping the check for resume runtime actually simplified this process so we can move the parsing into  DiagnosticsClientBuilder.Build.

lateralusX replied:
@noahfalk, meeting the bar, https://github.com/dotnet/diagnostics/pull/2374/commits/f194079ce6820db9b43ecc56404381afb8d07857?

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouterFactory.cs:
```diff
@@ -462,12 +462,21 @@ public async Task<Stream> AcceptIpcStreamAsync(CancellationToken token)
+                var path = Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}");
+                if (File.Exists(path))
+                {
+                    logger?.LogWarning($"Default IPC server path, {path}, already in use. To disable default diagnostics for dotnet-dsrouter, set COMPlus_EnableDiagnostics=0 and re-run.");
```
True, I will see if we can move it.

lateralusX replied:
Moved into away from dsrouter component into dotnet-dsrouter tool.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,126 @@
+                    }
+                }
+            }
+            catch (InvalidOperationException)
```
When parsing using UriKind.RelativeOrAbsolute it parse could succeed, but trying to access properties like Schema would throw InvalidOperationException if missing. Since then I switched this to UriKind.Absolute and then TryCreate is better to return failure if you don't have all needed components, so I believe we can drop this try/catch altogether.

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,126 @@
+            {
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && address.StartsWith(NamedPipeDefaultIPCRoot, StringComparison.OrdinalIgnoreCase))
```
There are some rules depending on the format when namedpipe, so believe that will still need to be done, but if I understand the above correct, using \.\pipe\ should be detected as a namedpipe regardless platform and trigger rules tied to format of namedpipe paths (like dropping part of the prefix) and once parsed, we could add logic enforcing what transports that are supported on specific OS:es, correct?

lateralusX replied:
@noahfalk, something like this, d5a9e34aa69c6818fe9271257cab442dd2eccfdd?

noahfalk replied:
> There are some rules depending on the format when namedpipe, so believe that will still need to be done
> @noahfalk, something like this, d5a9e34?

Yeah 👍 

I'm a little bummed we already introduced enough ambiguity that the same name can be interpretted as both a pipe or a UDS depending on the OS but I agree that preserving the existing behavior there is probably more useful than breaking it in the name of design purity.

josalem replied:
> I'm a little bummed we already introduced enough ambiguity that the same name can be interpretted as both a pipe or a UDS depending on the OS

I agree and regret not using a `<schema>://<address>,<tags>` syntax from the get-go. Would it be all that disruptive to modify both ends to use the more formalized/verbose style? Obviously, this goes beyond the scope of this PR, but it warrants a discussion. We're getting close to the point where every supported version of Windows will have Unix Domain Socket support. It's conceivable that we could add UDS support on Windows which would further muddy the waters if we don't adopt the more verbose style.

noahfalk replied:
> Would it be all that disruptive to modify both ends to use the more formalized/verbose style?

I'd suggest we leave it be until there is some immediate customer value to changing it.

> regret not using a <schema>://<address>,<tags>

We only need `\\.\pipe\blah` to eliminate the ambiguity. I expect our users still appreciate not having to write full URIs : )

lateralusX commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -48,6 +48,51 @@ internal abstract class IpcEndpoint
+
+        public static async Task<Stream> ConnectAsync(string address, CancellationToken token)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
This was mainly just moving existing code into a shared class, do you suggest we implement this now or do it as a separate PR at a later point? This class is internal, so we are free to continue extending on the transport support going forward and try to reduce the scope of this PR, OK?

noahfalk replied:
I'd prefer we resolved it in this PR because it appears like this PR is the one introducing the issue. When this code was in PidIpcEndpoint we were only reasoning about what the default transport was for a given OS and it was pretty closely localized with the code producing the remainder of the address string. Now that the code moved here we are expecting it correctly infer the transport for any address, including non-default ones. This PR also introduced the Parse code which is a new 2nd location that is duplicating the decision making about which transport is implied.

I didn't think this is a significant change. I was expecting this to be ~20-30 lines of code. Am I underestimating what it takes to fix this?

lateralusX replied:
I look into it.

josalem replied:
I wanted to echo Noah's point. Ideally, we'd have a single point where we parse/validate an input address and then internally we just pass around the resolved address which contains all the necessary information to disambiguate. Once we're past the parse/validate point, I'd prefer to not have strings getting passed around.

> PidEndpoint.GetDefaultAddress()

👍

lateralusX replied:
I implemented ability to pass around IpcEndpointConfig and use that in these methods, so now you can use an IpcEndpointConfig when creating DiagnosticClient and that will be used through the whole chain all the way down to the Connect/ConnectAsync methods.

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,156 @@
+        public enum TransportType
+        {
+            NamedPipe,
+            UnixDomainSocket
```
should this include TCP?

lateralusX replied:
I think it is better to add that once we have support for it.

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,156 @@
+                case TransportType.NamedPipe:
+                {
+                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        throw new PlatformNotSupportedException($"{NamedPipeSchema} only supported on Windows.");
```
```suggestion
                        throw new PlatformNotSupportedException($"{NamedPipeSchema} is only supported on Windows.");
```

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,156 @@
+                case TransportType.UnixDomainSocket:
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        throw new PlatformNotSupportedException($"{UnixDomainSocketSchema} not supported on Windows, use {NamedPipeSchema}.");
```
```suggestion
                        throw new PlatformNotSupportedException($"{UnixDomainSocketSchema} is not supported on Windows, use {NamedPipeSchema}.");
```

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,156 @@
+                address = parts[0];
+
+                if (parts.Length == 2)
+                {
+                    if (string.Equals(parts[1], "connect", StringComparison.OrdinalIgnoreCase))
+                    {
+                        portType = PortType.Connect;
+                    }
+                    else if (string.Equals(parts[1], "listen", StringComparison.OrdinalIgnoreCase))
+                    {
+                        portType = PortType.Listen;
+                    }
+                    else
+                    {
+                        throw new FormatException($"Unknow IPC endpoint config keyword, {parts[1]} in {config}.");
+                    }
```
```suggestion
                portType = parts[1].ToLower() switch
                {
                	"connect" => PortType.Connect,
                	"listen" => PortType.Listen,
                	_ => throw new FormatException($"Unknown IPC endpoint config keyword, {parts[1]} in {config}.");
                };
```

I'm a fan of the switch expression for this kind of stuff due to compactness, but it certainly isn't necessary to switch.

jander-msft replied:
If changing to switch, check that parts[1] is not null before using ToLower().

lateralusX replied:
I will keep the if/else if/else for now.

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,156 @@
+
+            if (Uri.TryCreate(address, UriKind.Absolute, out Uri parsedAddress))
+            {
+                if (string.Equals(parsedAddress.Scheme, NamedPipeSchema, StringComparison.OrdinalIgnoreCase))
+                {
+                    transportType = TransportType.NamedPipe;
+                    address = parsedAddress.AbsolutePath;
+                }
+                else if (string.Equals(parsedAddress.Scheme, UnixDomainSocketSchema, StringComparison.OrdinalIgnoreCase))
+                {
+                    transportType = TransportType.UnixDomainSocket;
+                    address = parsedAddress.AbsolutePath;
+                }
+                else if (!string.IsNullOrEmpty(parsedAddress.Scheme))
+                {
+                    throw new FormatException($"{parsedAddress.Scheme} not supported.");
+                }
```
```suggestion
                (transportType, address) = parsedAddress.Scheme.ToLower() switch
                {
                	NamedPipeSchema => (TransportType.NamedPipe, parsedAddress.AbsolutePath),
                	UnixDomainSocketSchema => (TransportType.UnixDomainSocket, parsedAddress.AbsolutePath),
                	_ => throw new FormatException($"{parsedAddress.Scheme} not supported.");
                };
```

Okay, I'll stop with the switch expression suggestions 😄

lateralusX replied:
I will keep the if/else if/else for now.

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs:
```diff
@@ -100,6 +145,51 @@ public override int GetHashCode()
+        public AddressIpcEndpoint(string address)
+        {
+            _address = address;
+        }
+
+        public override Stream Connect(TimeSpan timeout)
+        {
+            return IpcEndpointHelper.Connect(_address, timeout);
+        }
+
+        public override async Task<Stream> ConnectAsync(CancellationToken token)
+        {
+            return await IpcEndpointHelper.ConnectAsync(_address, token).ConfigureAwait(false);
+        }
+
+        public override void WaitForConnection(TimeSpan timeout)
+        {
+            using var _ = Connect(timeout);
+        }
+
+        public override async Task WaitForConnectionAsync(CancellationToken token)
+        {
+            using var _ = await ConnectAsync(token).ConfigureAwait(false);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as AddressIpcEndpoint);
+        }
+
+        public bool Equals(AddressIpcEndpoint other)
+        {
+            return other != null && other._address == _address;
+        }
+
+        public override int GetHashCode()
+        {
+            return _address.GetHashCode();
+        }
+    }
```
Rather than have an `AddressIpcEndpoint`, I think it might be a cleaner abstraction to have all `IpcEndpoint` variants have an `IpcEndpointConfig` on them. I should be able to ask an `IpcEndpoint` what its address and schema are.

lateralusX replied:
I have implemented something inline with this.

noahfalk replied:
> I should be able to ask an IpcEndpoint what its address and schema are

Sorry I didn't get to this earlier. While I am not opposed if you guys already implemented it, I don't understand what scenario we would ever invoke such code if we had it? For some our IpcEndpoints these values probably aren't well-defined, for example IpcEndpoints corresponding to processes that connect to listener we created.

lateralusX replied:
I only implemented the ability to use an IpcEndpointConfig, inline with comments elsewhere, so calls to Connect/ConnectAsync can take a more generic IpcEndpointConfig instead of string address.

josalem replied:
I'm good with that

josalem reviewed (commented):
A couple comments

lateralusX commented:
@noahfalk @jander-msft @josalem Thank you so much for all great feedback! I believe I have worked through most of them and implemented changes accordingly.

noahfalk commented:
> Thank you so much for all great feedback!

Np! Thanks for implementing it all : D

josalem commented on src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcEndpointConfig.cs:
```diff
@@ -0,0 +1,162 @@
+            return result != null;
+        }
+
+        public static IpcEndpointConfig Parse(string config)
```
Can we add a comment with valid examples of configs?

josalem reviewed (approved):
Took a look through your changes. LGTM :) thanks!
